src/pe/resource.rs

```diff
@@ -0,0 +1,1681 @@
+
+impl fmt::Debug for ResourceString<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut debug_struct = f.debug_struct("ResourceString");
+
+        debug_struct.field("len", &format_args!("{:#x}", self.len));
+
+        if self.is_text_data() {
+            // If this is string data, acrual size is multiple of 2 (sizeof wchar_t)
+            debug_struct.field(
+                "value_len",
+                &format_args!(
+                    "{:#x} ({} bytes)",
+                    self.value_len,
+                    self.value_len * SIZE_OF_WCHAR as u16
+                ),
+            );
+        } else {
+            debug_struct.field("value_len", &format_args!("{:#x}", self.value_len));
+        }
+
+        debug_struct
+            .field(
+                "type",
+                &format_args!(
+                    "{} ({})",
+                    self.r#type,
+                    match self.r#type {
+                        0 => "Binary Data",
+                        1 => "String Data",
+                        _ => "Unknown",
+                    }
+                ),
+            )
+            .field("key", &self.key_string())
+            .field(
+                "key_slice",
+                &format_args!("{:02x?} ({} bytes)", self.key, self.key.len()),
+            );
+
+        if self.is_text_data() && self.value_len > 0 {
+            debug_struct.field("value", &self.value_string());
+        }
+
+        debug_struct
+            .field(
+                "value_slice",
+                &format_args!(
+                    "{:02x?} ({} bytes, {})",
+                    self.value,
+                    self.value.len(),
+                    if self.value.len() == self.value_len as usize {
+                        "Correct"
+                    } else {
+                        "Incorrect"
+                    }
+                ),
+            )
+            .finish()
+    }
+}
+
+impl<'a> ResourceString<'a> {
+    pub fn parse(bytes: &'a [u8], offset: &mut usize) -> error::Result<Self> {
+        let len = bytes.gread_with::<u16>(offset, scroll::LE)?;
+        let value_len = bytes.gread_with::<u16>(offset, scroll::LE)?;
+        let r#type = bytes.gread_with::<u16>(offset, scroll::LE)?;
+        *offset = align_up(*offset, RESOURCE_STRING_FIELD_ALIGNMENT);
+        let key_size = &bytes[*offset..]
+            .chunks(2)
+            .take_while(|x| u16::from_le_bytes([x[0], x[1]]) != 0u16)
+            .count()
+            * SIZE_OF_WCHAR;
+        if (*offset - SIZE_OF_WCHAR) + key_size + SIZE_OF_WCHAR > bytes.len() {
+            return Err(error::Error::Malformed(format!(
+                "offset ({:#x}) and key_size ({:#x}) is greater than bytes len {:#x}",
+                offset,
+                key_size,
+                bytes.len()
+            )));
+        }
+        let key =
+            &bytes[*offset - SIZE_OF_WCHAR..*offset - SIZE_OF_WCHAR + key_size + SIZE_OF_WCHAR];
```

Comment:
this is a confusing offset computation, not sure it it can be simplified/extracted into subcomponents though :shrug: